src/components/tl/mlx5/tl_mlx5_pd.c

```diff
@@ -0,0 +1,293 @@
+{
+    struct sockaddr_storage sockaddr = {};
+    struct sockaddr_un *    addr;
+    int                     sock;
+    sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (sock == -1) {
+        tl_error(lib, "failed to create client socket errno %d", errno);
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    addr             = (struct sockaddr_un *)&sockaddr;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, sock_path, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) - 1] = '\0';
+
+    while (connect(sock, (struct sockaddr *)addr, SUN_LEN(addr)) == -1) {
+        if (errno != ENOENT) {
+            tl_error(lib, "failed to connect client socket errno %d", errno);
+            goto fail;
+        }
+    }
+    if (do_recvmsg(sock, shared_cmd_fd, shared_pd_handle) != UCC_OK) {
+        tl_error(lib, "Failed to recv msg");
+        goto fail;
+    }
+
+    if (close(sock) == -1) {
+        tl_error(lib, "Failed to close client socket errno %d", errno);
+        return UCC_ERR_NO_MESSAGE;
+    }
+    return UCC_OK;
+
+fail:
+    if (close(sock) == -1) {
+        tl_error(lib, "Failed to close client socket errno %d", errno);
+    }
+    return UCC_ERR_NO_MESSAGE;
+}
+
+static ucc_status_t server_send_data(int command_fd, uint32_t pd_handle,
+                                     int group_size, int sock,
+                                     ucc_tl_mlx5_lib_t *lib)
+{
+    int                i = 0;
+    connection_t       connection[group_size];
+    pthread_t          thread[group_size];
+    struct sockaddr_un addr;
+    socklen_t          addrlen;
+
+    while (i < group_size) {
+        /* accept incoming connections */
+        connection[i].sock      = accept(sock, NULL, 0);
+        connection[i].fd        = command_fd;
+        connection[i].pd_handle = pd_handle;
+        if (connection[i].sock != -1) {
+            /* start a new thread but do not wait for it */
+            pthread_create(&thread[i], 0, do_sendmsg, (void *)&connection[i]);
+            i++;
+        }
+    }
+
+    for (i = 0; i < group_size; i++) {
+        pthread_join(thread[i], NULL);
+    }
+
+    addrlen = sizeof(addr);
+    getsockname(sock, (struct sockaddr *)&addr, &addrlen);
+
+    for (i = 0; i < group_size; i++) {
+        if (connection[i].return_val != UCC_OK) {
+            tl_error(lib, "failed to send cmd_fd");
+            goto listen_fail;
+        }
+    }
+
+    if (close(sock) == -1) {
+        tl_error(lib, "failed to close server socket errno %d", errno);
+        if (remove(addr.sun_path) == -1) {
+            tl_error(lib, "socket file removal failed");
+        }
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    if (remove(addr.sun_path) == -1) {
+        tl_error(lib, "socket file removal failed");
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    return UCC_OK;
+
+listen_fail:
+    if (remove(addr.sun_path) == -1) {
+        tl_error(lib, "socket file removal failed");
+    }
+    if (close(sock) == -1) {
```

Comment:
is sock close should before removing sun_path?

Reply:
sounds right